File: trois_couleurs.py
import cProfile
import numpy as np
from utillc import *
import sys
import sympy
from sympy import symbols, lambdify, Integer, python
import matplotlib.pyplot as plt 
from typing import TypeVar
import numpy as np
from numpy.typing import NDArray
import numpy
from numba import jit
from numba import njit
from itertools import chain

#from numba.typed import List as List
List = lambda x : list(x)

from typing import Literal, Tuple, TypeVar
# Generic dimension sizes types
T1 = TypeVar("T1", bound=int)
T2 = TypeVar("T2", bound=int)
T3 = TypeVar("T3", bound=int)

# Dimension types represented as typles
Shape = Tuple
Shape1D = Shape[T1]
Shape2D = Shape[T1, T2]
Shape3D = Shape[T1, T2, T3]

ROWS = Literal[2]
COLS = Literal[3]
A: NDArray[Shape2D[ROWS, COLS]] = np.zeros((2,3))
B: NDArray[Shape2D[COLS, ROWS]] = np.zeros((3,2))
C: NDArray[Shape2D[ROWS, ROWS]] = np.dot(A, B)


utillc.print_everything()


import inspect, os

"""
 rotation les couleurs

crot
# 0 => 0
# 1 => 2
# 2 => 3
# 3 => 1

et
 crot2
# 0 => 0
# 1 => 1
# 2 => 3
# 3 => 2

 on a tout par : x, crot(x), crot(crot(x)), crot2(x), crot(crot2(x)), crot(crot(crot2(x)))

"""

def rot_func1(t) :
    """
    calcul d'une rotation des couleurs en utilisant un polynome
    p(x) = a + b x + c x*x + d * x*x*x
    p(0) -> 0 - c'est le codage d'une case vide qui doit rester vide
    p(1) -> p1, p(2) -> p2, p(3) -> p3
    on determine les coefs en fournissant les pi ( dans le param t)
    """
    a,b,c,d,x = symbols("a,b,c,d,x", real=True)
    pol = symbols("pol", integer=True)
    p1, p2, p3 = t
    eq = []
    eq += [ a ]
    eq += [ a + b   + c   + d        - p1] 
    eq += [ a + 2*b + 4*c + 8*d      - p2] 
    eq += [ a + 3*b + 9*c + 3*3*3*d  - p3]
    
    p = a + b*x + c*x*x + d*x*x*x
    s = sympy.solve(eq, [a,b,c,d])
    #EKOX(s)
    #EKOX(p.subs(s))
    #f1 = lambdify([ a, b, c, d], s, "numpy")

    f1 = lambdify([ x ], p.subs(s), "numpy")
    
    #EKOX(t)
    #EKOX(f1(1))
    #EKOX(f1(2))
    #EKOX(f1(3))
    
    f1j = jit(f1, nopython=True)
    return f1j, f1, p.subs(s)

def rot_func(x) :
    return rot_func1(x)[0]


#EKOX(python(rot_func1((2, 3, 1))[2]))
#EKOX(python(rot_func1((1, 3, 2))[2]))



# pour construire toutes les variantes équivalentes en couleur d'une position
#EKO()
crotf_ =  rot_func((2, 3, 1))
crot2f_ = rot_func((1, 3, 2))

# dump du résultat de sympy
#@jit(nopython=True)
def crotf_(x) -> np.float64 : return -x**3/3 + x**2/2 + 11*x/6

#@jit(nopython=True)
def crot2f_(x) -> np.float64 :return -2*x**3/3 + 5*x**2/2 - 5*x/6

#@jit(nopython=True)
def crotf(x) -> np.uint8 :  return np.round(crotf_(x)).astype(np.uint8) 

#@jit(nopython=True)
def crot2f(x)  -> np.uint8 : return np.round(crot2f_(x)).astype(np.uint8)

ar = lambda x : np.asarray(x)
aa = ar((0, 1,2,3))


"""
EKOX(aa)
EKOX(crotf(aa))
EKOX(crotf(crotf(aa)))
EKOX(crot2f(aa))
EKOX(crotf(crot2f(aa)))
EKOX(crot2f(crotf(aa)))

EKON([ crotf(x) for x in range(4)])
EKOX([ crot2f(x) for x in range(4)])
"""


class Game :

    def __init__(self, DDD=2, opt1=False, reverse=True) :
        self.node_number = 0
        self.reverse = reverse
        DN,DM = DDD, DDD # DN == DM car le calcul de symetries l'obligent
        DN_ = Literal[DN]
        DM_ = Literal[DM]
        self.taille_plateau = (DN, DM)
        assert(DN== DM) # on veut un tableau carré because, on gere pas les rotations autrement

        # la taille d'un plateau : N x N
        self.TT = NDArray[Shape2D[DN_, DM_]]
        self.DN, self.DM = DN, DM
        self.step_count = 0
        self.opt1 = opt1
        
    def inc(self) :
        x = self.node_number
        self.node_number += 1
        return x

    def go(self, DDD = 2) :
        #DDD = 2
        TT = self.TT
        DN, DM = self.DN, self.DM
        game= self
        class Node :
            """
            les noeuds de l'arbre
            coords dans le tableau: y,x
            """

            def __init__(self, b: TT, father = None) :
                self.number = game.inc()
                self.board = b
                self.losing = "?" # pour celui qui doit jouer dans cette position
                self.father = father

                # descendant, order == Node creation order 
                self.children = []

                # les noeuds équivalents sont chainés entre eux
                self.friend = None

                if father is not None :
                    father.children.append(self)

            def name(self) :
                return "n" + str(self.number) 


            def chain(self, ref) :
                assert(self.friend == None)
                self.friend = ref

            def empty(self) -> NDArray :
                res = np.argwhere(self.board == 0)
                return res

            def neighbours(self, c : NDArray, filled = True) -> list:
                """
                cases voisines de c
                """
                y, x = c
                l = [ (y, x+1), (y, x-1), (y+1, x), (y-1, x) ]

                if filled :
                    l = [ (yy,xx) for (yy, xx) in l if
                          xx >= 0 and xx < DM and yy >= 0 and yy < DN and self.board[yy, xx] > 0]
                else :
                    l = [ (yy,xx) for (yy, xx) in l if
                          xx >= 0 and xx < DM and yy >= 0 and yy < DN and self.board[yy, xx] == 0]
                return l

            def possible_moves(self) -> list :
                """
                les positions atteignables à partir de celle là
                """
                e = self.empty()
                res = []
                for c in e :
                    nb = self.neighbours(c, filled = True)
                    nb = np.asarray(nb).T
                    if len(nb) == 0 :
                        colors_available = colors
                    else :
                        colors_around = set(self.board[tuple(nb)])
                        colors_available = colors - colors_around
                    for e in colors_available : res += [ ( c, e, len(nb)) ]
                if game.opt1 :
                    res_sorted = sorted(res, key = lambda ee : ee[2], reverse=game.reverse)
                    return [ (e[0], e[1]) for e in res_sorted]
                else :
                    return [ (e[0], e[1]) for e in res]

            def eq(self, nn) :
                return nn == ''.join(map(str,self.board.flatten()))

            def level(self) :
                return DN*DM - (self.board == 0).sum()

            def apply(self, m : tuple) -> np.array :
                (y, x), col = m
                b = self.board.copy()
                b[y, x] = col
                return b

        EKON(self.taille_plateau)

        # coding cells
        factors = [ 4**i for i in range(DN*DM)]
        if False :
            EKOX(factors[-1])
            EKOX(pow(factors[-1], 1/(DN*DM - 1)))
            EKOX(('factors', factors))
        factors = np.asarray(factors).reshape((DN, DM))
        #EKOX(factors.shape)
        #EKOX(('factors', factors))

        # codage des couleurs
        # 0 : vide, 1, 2, 3 les 3 couleurs

        colors = set([1,2,3])

        # codage d'une position : le tableau => un nombre
        #@jit(nopython=True)
        def hh(b : TT) -> numpy.int64 :
            """ codage d'une position : le tableau => un nombre    
            """
            xx = b * factors
            return xx.sum()


        # random board
        def random() -> TT :
            """
            random board
            """
            b = np.random.uniform(0, 3, size=(DN, DM)).round().astype(np.uint8)
            return b

        aa = random()
        if False :
            EKOX(hh(aa))
            EKOX(hh(crotf(aa)))
            EKOX(hh(crotf(crotf(aa))))
            EKOX(hh(crot2f(aa)))
            EKOX(hh(crotf(crot2f(aa))))
            EKOX(hh(crot2f(crotf(aa))))

        #board vide
        def zero() -> TT :
            """ empty board
            """
            b = np.zeros(shape = (DN, DM)).astype(np.uint8)
            return b


        @jit(nopython=True)
        def mychain(l : List) -> List:
            # chain('ABC', 'DEF') → A B C D E F
            for iterable in l:
                yield from iterable

        #@jit(nopython=True)
        def normal(bb : TT) -> numpy.int64:
            """
            calcule une représentation unique quelquesoit les variantes qui sont équivalentes
            => par rotation et symetries
            => par changement de couleur
            """
            emp = (bb == 0).sum()
            if emp == DN*DM :
                return 0

            b = bb
            b1 = np.flipud(b); 
            b2 = np.fliplr(b); 
            b3 = np.rot90(b); 
            b4 = np.rot90(b3);
            b5 = np.rot90(b4); 
            b6 = b.T; 
            b7 = b[::-1, ::-1].T; 

            l = List([ b, b1, b2, b3,  b4,  b5, b6, b7])
            l0 = l
            l1 = map(crotf, l)
            l2 = map(crotf, l1)
            l3 = map(crot2f, l)
            l4 = map(crotf, l3)
            l5 = map(crot2f, l1)

            ss = map(hh, chain(l0, l1, l2, l3, l4, l5))

            #ss = map(hh, mychain((l0, l1, l2, l3, l4, l5)))
            #ss = map(hh, l0)
            #ss = map(hh, l1)
            #ss = map(hh, l2)
            #ss = map(hh, l0+l1)
            #ss = map(hh, [ e for ee in (l0, l1, l2) for e in ee])
            #ss = map(hh, [ e for ee in (l0, l1, l2, l3, l4, l5) for e in ee])    
            ss = sorted(ss)
            return ss[0]

        """
        aa = np.asarray([[ 1,2,3,1],
                         [ 1,0,0,0],
                         [ 1,1,0,0],
                         [ 1,0,0,0]])
        EKOX(normal(aa))
        aa = np.asarray([[ 1,2,3,2],
                         [ 1,0,0,0],
                         [ 1,1,1,0],
                         [ 1,0,0,0]])
        EKOX(normal(aa))
        aan = Node(aa)
        ms = aan.possible_moves()
        nexts = [ aan.apply(m) for m in ms]
        #EKOX(nexts)
        """



        max_front = 1
        min_empty_cells = DN*DM * 12

        cc=[]

        def parse(r, tab) :
            if normal(r) in seen :
                r = seen[normal(r)]

            p = ('\n' + tab + 'losing=' + str(r.losing) +
                 '\n' + tab + 'level=' + str(r.level()) +    
                 '\n' + tab + str(r.board[0]) +
                 '\n' + tab + str(r.board[1]))
            ss = ''.join([ parse(c, tab + '\t') for c in r.children])
            return p + ss



        def dump(n, t="") :
            EKON(normal(n.board), n.level())
            for e in n.children :
                dump(e, t+"\t")

        def check(r, father) :
            """
            sanity check of the built tree
            """
            if father is not None :
                assert(father.level() == r.level() - 1)
            #EKOX([( r.level(), c.level()) for c in r.children])
            [ check(c, r) for c in r.children]    



        def plus() :
            self.step_count += 1
            return self.step_count


        def recursive_build(p : Node) :
            """
            recursive prog, donc profondeur d'abord
            """
            istep = plus()
            nn = normal(p.board)

            empy_cells_num = (p.board == 0).sum()
            #EKON(empy_cells_num, min_empty_cells)
            if p.eq("1201") :    EKO()
            #EKOX(step)
            #EKOX(parse(root, ""))
            #check(root, None)

            if node_number in  [ 2**i for i in range(1,30)] :
                EKOX(node_number)

            if empy_cells_num >= 0 :
                if nn in seen :
                    assert(seen[nn].level() == p.level())
                    # on a déjà traité un noeud de clé identique : q = seen[nn]
                    # pas besoin de poursuivre
                    p.chain(ref = seen[nn]);
                else :
                    seen[nn] = p            
                    ms = p.possible_moves()
                    if len(ms) == 0 :
                        p.losing = True
                    else :
                        ls = []
                        p.losing = True
                        for m in ms :
                            new_node  = Node(p.apply(m), p)
                            recursive_build(new_node)
                            if new_node.losing :
                                # alpha beta pruning
                                p.losing = False
                                break
                            ls.append(new_node)
            else :
                # terminal node
                EKON(min_empty_cells, len(seen), nn, istep)
                EKOX(p.board)

        def solve(r) :
            """
            assuming depth first stat (or recursive), friend must have been visited yet
            """
            if r.friend is not None :
                assert(r.number > r.friend.number)
                r.losing = r.friend.losing
            else :
                for c in r.children :
                    assert(r.level() + 1 == c.level())
                    solve(c)
                    # no children => empty list => True
                # 'relative' losing == the gain for the player in this position
                # i win if one among next positions is losing for the other player 
                # i loose if all  next positions is wining for the other player 
                r.losing = all([ (not c.losing) for c in r.children])

        node_number, root, seen = 0, Node(zero()), {}

        #EKOT("building recursif...")
        recursive_build(root)
        #EKON(len(seen))
        #EKOX(self.step_count)
        EKOX(self.node_number)
        solve(root)
        EKOX(root.losing)



        def dot(r, fd) :
            status = "L" if r.losing else "W"
            nrm = normal(r.board)
            ref = "ref:" + seen[nrm].name() if nrm in seen else ""
            board = str(r.board)
            bb = ""
            fd.write(r.name() + ' [ label="' + r.name() + " " + status + ' \\n ' + bb + " " + ref + '"]')
            for e in r.children :
                fd.write("n" + str(r.number) + " -> n" + str(e.number) + ";\n")
                dot(e, fd)


        if False :
            with open('dot.dot', 'w') as f:
                f.write("digraph troiscouleurs {")
                dot(root, f)
                f.write("}")

for nn in range(1, 5) :
    game = Game(nn)
    game.go()
for nn in range(1, 5) :
    game = Game(nn, opt1=True, reverse=True)
    game.go()
for nn in range(1, 5) :
    game = Game(nn, opt1=True, reverse=False)
    game.go()

File: trois_couleurs.js

//math = require("mathjs")

import * as math from 'mathjs';

const List = (x) => Array.from(x);

// Generic dimension sizes types
const ROWS = 2;
const COLS = 3;
const A = Array.from({ length: ROWS }, () => Array(COLS).fill(0));
const B = Array.from({ length: COLS }, () => Array(ROWS).fill(0));
const C = math.multiply(A, B);

// Placeholder for utillc.print_everything()
console.log("Print everything function called");

function rot_func(t) {
    const [p1, p2, p3] = t;
    const a = math.symbol('a');
    const b = math.symbol('b');
    const c = math.symbol('c');
    const d = math.symbol('d');
    const x = math.symbol('x');
    
    const eq = [
        a,
        math.add(math.add(math.add(a, b), c), d).subtract(p1),
        math.add(math.add(math.add(a, math.multiply(2, b)), math.multiply(4, c)), math.multiply(8, d)).subtract(p2),
        math.add(math.add(math.add(a, math.multiply(3, b)), math.multiply(9, c)), math.multiply(27, d)).subtract(p3)
    ];
    
    const p = math.add(a, math.add(math.multiply(b, x), math.add(math.multiply(c, math.pow(x, 2)), math.multiply(d, math.pow(x, 3)))));
    const s = math.solve(eq, [a, b, c, d]);
    
    const f1 = (x) => p.subs(s);
    
    return f1;
}

const crotf_ = rot_func([2, 3, 1]);
const crot2f_ = rot_func([1, 3, 2]);
const crotf = (x) => Math.round(crotf_(x)).toString(2);
const crot2f = (x) => Math.round(crot2f_(x)).toString(2);

const ar = (x) => Array.from(x);
const aa1 = ar([0, 1, 2, 3]);

const DDD = 4;
const DN = DDD, DM = DDD; // DN == DM car le calcul de symetries l'obligent
const taille_plateau = [DN, DM];
console.assert(DN === DM); // on veut un tableau carré because, on gere pas les rotations autrement

const TT = Array.from({ length: DN }, () => Array(DM).fill(0));
console.log("EKON called with taille_plateau");

let factors = Array.from({ length: DN * DM }, (_, i) => Math.pow(4, i));
factors = Array.from({ length: DN }, (_, i) => factors.slice(i * DM, (i + 1) * DM));

// codage des couleurs
const colors = new Set([1, 2, 3]);

function hh(b) {
    const xx = b.map((row, i) => row.map((val, j) => val * factors[i][j])).flat();
    return xx.reduce((acc, val) => acc + val, 0);
}

function random() {
    const b = Array.from({ length: DN }, () => Array.from({ length: DM }, () => Math.round(Math.random() * 3)));
    return b;
}

const aa = random();

function zero() {
    return Array.from({ length: DN }, () => Array(DM).fill(0));
}

function normal(bb) {
    const emp = bb.flat().filter(x => x === 0).length;
    if (emp === DN * DM) {
        return 0;
    }
    
    const b = bb;
    const b1 = b.map(row => row.slice().reverse());
    const b2 = b.map(row => row.reverse());
    const b3 = b[0].map((_, colIndex) => b.map(row => row[colIndex])).reverse();
    const b4 = b3.map(row => row.slice().reverse());
    const b5 = b4.map(row => row.slice().reverse());
    const b6 = b[0].map((_, colIndex) => b.map(row => row[colIndex]));
    const b7 = b.map(row => row.slice().reverse()).map(row => row.slice().reverse());
    
    const l = List([b, b1, b2, b3, b4, b5, b6, b7]);
    const l0 = l;
    const l1 = l.map(crotf);
    const l2 = l1.map(crotf);
    const l3 = l.map(crot2f);
    const l4 = l3.map(crotf);
    const l5 = l1.map(crot2f);
    const ss = [...l0, ...l1, ...l2, ...l3, ...l4, ...l5].map(hh).sort();
    return ss[0];
}

var node_number = 0;
function inc() {
    return node_number++;
}

class Node {
    constructor(b, father = null) {
        this.number = inc();
        this.board = b;
        this.losing = "?";
        this.father = father;
        this.children = [];
        this.friend = null;
        
        if (father !== null) {
            father.children.push(this);
        }
    }

    name() {
        return "n" + this.number;
    }

    chain(ref) {
        console.assert(this.friend === null);
        this.friend = ref;
    }

    empty() {
        return this.board.flatMap((row, y) => row.map((val, x) => (val === 0 ? [y, x] : null))).filter(Boolean);
    }

    neighbours(c, filled = true) {
        const [y, x] = c;
        let l = [[y, x + 1], [y, x - 1], [y + 1, x], [y - 1, x]];

        if (filled) {
            l = l.filter(([yy, xx]) => xx >= 0 && xx < DM && yy >= 0 && yy < DN && this.board[yy][xx] > 0);
        } else {
            l = l.filter(([yy, xx]) => xx >= 0 && xx < DM && yy >= 0 && yy < DN && this.board[yy][xx] === 0);
        }
        return l;
    }

    possible_moves() {
        const e = this.empty();
        const res = [];
        for (const c of e) {
            const nb = this.neighbours(c, true);
            const nbTransposed = nb[0].map((_, colIndex) => nb.map(row => row[colIndex]));
            const colors_available = nbTransposed.length === 0 ? colors : new Set([...colors].filter(color => !new Set(nbTransposed.map(([yy, xx]) => this.board[yy][xx])).has(color)));
            for (const e of colors_available) res.push([c, e]);
        }
        return res;
    }

    eq(nn) {
        return nn === this.board.flat().join('');
    }

    level() {
        return DN * DM - this.board.flat().filter(x => x === 0).length;
    }

    apply(m) {
        const [[y, x], col] = m;
        const b = this.board.map(row => row.slice());
        b[y][x] = col;
        return b;
    }
}

let max_front = 1;
let min_empty_cells = DN * DM * 12;

let cc = [];

function parse(r, tab) {
    if (normal(r) in seen) {
        r = seen[normal(r)];
    }

    const p = `\n${tab}losing=${r.losing}\n${tab}level=${r.level()}\n${tab}${r.board[0]}\n${tab}${r.board[1]}`;
    const ss = r.children.map(c => parse(c, tab + '\t')).join('');
    return p + ss;
}

function dump(n, t = "") {
    console.log(normal(n.board), n.level());
    for (const e of n.children) {
        dump(e, t + "\t");
    }
}

function check(r, father) {
    if (father !== null) {
        console.assert(father.level() === r.level() - 1);
    }
    r.children.forEach(c => check(c, r));
}

let step_count = 0;
function plus() {
    return step_count++;
}

function recursive_build(p) {
    const istep = plus();
    const nn = normal(p.board);

    const empy_cells_num = p.board.flat().filter(x => x === 0).length;

    if (empy_cells_num >= 0) {
        if (nn in seen) {
            console.assert(seen[nn].level() === p.level());
            p.chain(seen[nn]);
        } else {
            seen[nn] = p;
            const ms = p.possible_moves();
            if (ms.length === 0) {
                p.losing = true;
            } else {
                p.losing = true;
                for (const m of ms) {
                    const new_node = new Node(p.apply(m), p);
                    recursive_build(new_node);
                    if (new_node.losing) {
                        p.losing = false;
                        break;
                    }
                }
            }
        }
    } else {
        console.log(min_empty_cells, seen.length, nn, istep);
        console.log(p.board);
    }
}

function solve(r) {
    if (r.friend !== null) {
        console.assert(r.number > r.friend.number);
        r.losing = r.friend.losing;
    } else {
        for (const c of r.children) {
            console.assert(r.level() + 1 === c.level());
            solve(c);
        }
        r.losing = r.children.every(c => !c.losing);
    }
}

node_number = 0;
const root = new Node(zero());
const seen = {};

console.log("building recursif...");
recursive_build(root);
console.log(seen.length);
console.log(step_count);
console.log(node_number);
solve(root);
console.log(root.losing);
